pkg/scheduler/metrics/profile_metrics.go

```diff
@@ -0,0 +1,56 @@
+package metrics
+
+import (
+	"time"
+
+	"k8s.io/component-base/metrics"
+)
+
+var (
+	scheduledResult     = "scheduled"
+	unschedulableResult = "unschedulable"
+	errorResult         = "error"
+)
+
+type PerProfile struct {
+	perResult map[string]perResult
+}
```

Comment:
I feel like we're complicating the case. Isn't Prometheus maintaining the map for us?

As `scheduleAttempts` has already been registered, I think we can directly use:

```go
// "result value" = scheduled | unschedulable | error
// "profile name" is dynamic 
 scheduleAttempts.With(metrics.Labels{"result": "<result value>", "profile": "<profile name>"})`
```

in runtime to get the Counter, and then record the metrics properly.

(not a Prometheus expert, needs @logicalhan to confirm. Maybe there is a better way to achieve this.)

Reply:
https://godoc.org/github.com/prometheus/client_golang/prometheus#CounterVec.GetMetricWithLabelValues

`Keeping the Counter for later use is possible (and should be considered if performance is critical), but keep in mind that Reset, DeleteLabelValues and Delete can be used to delete the Counter from the CounterVec. In that case, the Counter will still exist, but it will not be exported anymore, even if a Counter with the same label values is created later.`

There seems to be validation and hashing happening. I think it's fair to keep them, given that they are going to be used so much. I'm even considering using a slice instead of the map, but maybe that's overkill.

Reply:
Agree with Aldo, I think it is likely better to get references to the counters at the beginning. We just need to make sure they are properly initialized before they get used.

Reply:
> There seems to be validation and hashing happening. I think it's fair to keep them, given that they are going to be used so much.

That's fair, given that we **know** the particular profiles we'll use.

Reply:
You can actually invoke: `WithLabelValues(lvs ...string)` (where you need the label values to map to the order of the declared metric labels) or `With(labels map[string]string)`.